### PR TITLE
Add constant weight representation in Keras & PyTorch

### PR DESCRIPTION
**Pull request by elad-c**
## Pull Request Description:
This PR refactors the way MCT handles constants. Instead of keeping them in the framework_attr, the constants are saved as weights in the internal graph nodes. The constants are referred to as "positional weights" whos keys are integers corresponding to the position of the constants in the call_args\call_kwargs. The constants are restored in the model builder.
This PR is the first step towards fully quantizing all constants in a model

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/keras/back2framework/keras_model_builder.py [hunk 1]
```diff
@@ -270,6 +270,7 @@ def _run_operation(self,
                                                                            out_tensors_of_n_float)
         else:
             input_tensors = [tensor for tensor_list in input_tensors for tensor in tensor_list]  # flat list of lists
+            input_tensors = n.insert_positional_weights_to_input_list(input_tensors)
```
Comment (lapid92) [inline model_compression_toolkit/core/keras/back2framework/keras_model_builder.py:273]: Maybe explain what is positional weights

**Review by lapid92 — APPROVED** [review]

File: tests/pytorch_tests/model_tests/base_pytorch_test.py [hunk 2]
```diff
@@ -33,7 +33,7 @@
 class BasePytorchTest(BaseFeatureNetworkTest):
     def __init__(self,
                  unit_test,
-                 float_reconstruction_error=1e-7,
+                 float_reconstruction_error=1e-6,
```
Comment (reuvenperetz) [inline tests/pytorch_tests/model_tests/base_pytorch_test.py:36]: Why this change is needed?

File: tests/keras_tests/feature_networks_tests/feature_networks/const_representation_test.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        x3 = tf.add_n([x1, np.random.random(x.shape.as_list()).astype(np.float32), x2])
+        x = tf.concat([x1, x2, np.random.random(x3.shape.as_list()).astype(np.float32), x3], 1)
+        return tf.keras.models.Model(inputs=inputs, outputs=x)
+
+    def compare(self, quantized_model, float_model, input_x=None, quantization_info=None):
+        y = float_model.predict(input_x)
+        y_hat = quantized_model.predict(input_x)
+        self.unit_test.assertTrue(y.shape == y_hat.shape, msg=f'out shape is not as expected!')
+        cs = cosine_similarity(y, y_hat)
+        self.unit_test.assertTrue(np.isclose(cs, 1), msg=f'fail cosine similarity check:{cs}')
```
Comment (reuvenperetz) [inline tests/keras_tests/feature_networks_tests/feature_networks/const_representation_test.py:111]: I can live the CS comparison since this model is not quantized, but can we test additionally that the constants are quantized as they should?

File: tests/pytorch_tests/model_tests/feature_models/bn_function_test.py [hunk 4]
```diff
@@ -18,7 +18,7 @@
 from model_compression_toolkit.core.pytorch.pytorch_device_config import get_working_device
 
 """
-This tests check the batch_norm function and demonstrates the usage of BufferHolder node.
+This tests check the batch_norm function.
```
Comment (lior-dikstein) [inline tests/pytorch_tests/model_tests/feature_models/bn_function_test.py:21]: These

File: tests/pytorch_tests/model_tests/feature_models/bn_function_test.py [hunk 5]
```diff
@@ -39,7 +39,7 @@ def forward(self, x, y):
 
 class BNFNetTest(BasePytorchTest):
     """
-    This tests check the batch_norm function and demonstrates the usage of BufferHolder node.
+    This test check the batch_norm function.
```
Comment (lior-dikstein) [inline tests/pytorch_tests/model_tests/feature_models/bn_function_test.py:42]: checks

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 6]
```diff
@@ -191,7 +194,7 @@ def test_second_moment_correction(self):
 
     def test_bn_function(self):
         """
-        This tests check the batch_norm function and demonstrates the usage of BufferHolder node.
+        This test check the batch_norm function.
```
Comment (lior-dikstein) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:197]: checks

File: model_compression_toolkit/core/common/graph/base_node.py [hunk 7]
```diff
@@ -180,7 +180,23 @@ def get_weights_list(self):
         Returns: A list of all weights the node holds.
 
         """
-        return [self.weights[k] for k in self.weights.keys() if self.weights[k] is not None]
+        return [self.weights[k] for k in self.weights.keys()
+                if self.weights[k] is not None and not isinstance(k, int)]
+
+    def insert_positional_weights_to_input_list(self, input_tensors: List) -> List:
```
[the remark this answers is not in the document]
Comment (elad-c) [reply model_compression_toolkit/core/common/graph/base_node.py:186]: it's framework specific Tensor object, so can't type hint it as this is in common

File: tests/keras_tests/feature_networks_tests/feature_networks/const_representation_test.py (hunk 3 above)
Comment (elad-c) [reply tests/keras_tests/feature_networks_tests/feature_networks/const_representation_test.py:111]: quantization of positional weights is the next PR

File: model_compression_toolkit/core/common/graph/base_node.py [hunk 8]
```diff
@@ -177,10 +179,28 @@ def set_weights_by_keys(self, name: str, tensor: np.ndarray):
...
+        in the node's list of inputs according to their keys in the weights dictionary.
+
+        Args:
+            input_tensors: activation input tensors to node.
+        Returns:
+            Activation input tensors list with positional weights
+        """
+        for pos, weight in sorted((pos, weight) for pos, weight in self.weights.items()
+                                  if isinstance(pos, int)):
+            assert pos <= len(input_tensors), 'Positional weight index mismatch'
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/graph/base_node.py:200]: Please use logger

**Review by reuvenperetz — APPROVED** [review]
